sw/device/silicon_creator/rom/doc/rom_customization.md

```diff
@@ -0,0 +1,98 @@
+          └──────┬─────┘
+ ROM   ┌─────────┴─────────┐  ROM
+┌State─▽─────┐      ┌──────▽──State
+│ ┌────────┐ │      │ ┌────────┐ │
+│ │Pre-run │ │      │ │Pre-run │ │
+│ ├────────┤ │      │ ├────────┤ │
+│ │  Run   │ │      │ │  Run   │ │
+│ ├────────┤ │      │ ├────────┤ │
+│ │Post-run│ │      │ │Post-run│ │
+│ └────────┘ │      │ └────────┘ │
+└────────────┘      └──────┬─────┘
+                           │  ROM
+                    ┌──────▽──State
+                    │ ┌────────┐ │
+                    │ │Pre-run │ │
+                    │ ├────────┤ │
+                    │ │  Run   │ │
+                    │ ├────────┤ │
+                    │ │Post-run│ │
+                    │ └────────┘ │
+                    └────────────┘
+```
+
+With this proposal, vendors can customize an OpenTitan ROM by providing custom implementations for the pre-run and post-run hooks of any ROM state. By default, these hooks are empty.
+
+The main ROM state implementation, the run callback, is executed in step 2. All ROM state run callbacks are defined by the upstream ROM implementation and are not customizable.
+
+Crucially, only the run callback determines the next state in the ROM state machine. Pre-run and post-run hooks cannot override these transitions, preventing vendors from directly modifying the ROM's execution flow.
+
+In essence, this proposal enables vendors to customize the execution of individual ROM states, but it strictly controls the overall ROM execution flow.
+
+### ROM states
+
+Each OpenTitan ROM defines a set of execution states and associates each state with a specific ROM state run callback. A ROM state run callback takes an opaque pointer argument (which is shared with the state hooks) and returns the next state in the sequence.
+
+To enable effective vendor customization, ROM states should be clearly defined and map to specific functional steps in the ROM's execution flow. For instance, `kRomStateBootstrap` is a well-defined state, while `kRomStateStep3` is too generic.
+
+The more precisely a ROM state is defined, the easier it is for vendors to understand and customize its behavior.
+
+### ROM state hooks
+
+Pre-run and post-run ROM state hooks are executed by the ROM state machine before and after each state run callback, respectively. They share state with the state run callbacks through an opaque pointer argument.
+
+Unlike the state run callback, ROM state hooks cannot determine the next state in the ROM state machine. Vendors can customize specific ROM state execution steps through these hooks by providing custom implementations in an external repository.
+
+Similar to the existing manufacturing and test hook customization framework, custom ROM state hooks can be included in a ROM image by specifying the path to the vendor hook implementations in an environment variable passed to the ROM build system:
+
+``` bash
+ROM_HOOKS_DIR=/PATH/TO/LOCAL/ROM/HOOKS ./bazelisk.sh build //sw/device/silicon_creator/rom:mask_rom
```

Comment:
Comment from @a-will in https://github.com/lowRISC/opentitan/issues/25633#issuecomment-2549077273:

Just an implementation nit: It would be nice if we could avoid more environment variable dependencies. Perhaps bazel's --override_module could help here instead, where we define a rom_hooks module (which contains the upstream implementations) that could be overridden.